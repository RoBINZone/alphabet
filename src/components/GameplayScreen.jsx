import React, { useEffect } from "react";
import { StatisticsService } from "../services/statistics-service.js";
import {
  LETTERS,
  MAXTURNS,
  SET_SCREEN,
  SET_SELECTED_LETTERS_INDEXES,
  STATISTICS_SCREEN,
} from "../consts.js";
import { GameplayService } from "../services/gameplay-service.js";
import { GameSelectors } from "../selectors/game-selectors.js";
import { GameActions } from "../actions/game-actions.js";

export const GameplayScreen = ({ gameState, dispatch }) => {
  useEffect(() => {
    dispatch({
      type: SET_SELECTED_LETTERS_INDEXES,
      selectedLettersIndexes: GameplayService.get2RandomLetters(),
    });
  }, [dispatch]);

  useEffect(() => {
    if (gameState?.answers.length >= MAXTURNS) {
      StatisticsService.setStatistic(
        gameState.startTime,
        GameSelectors.getCorrectCount(gameState),
        GameSelectors.getIncorrectCount(gameState),
        GameSelectors.getAverageResponseTime(gameState),
      );
      dispatch({ type: SET_SCREEN, screen: STATISTICS_SCREEN });
    } else {
      dispatch({
        type: SET_SELECTED_LETTERS_INDEXES,
        selectedLettersIndexes: GameplayService.get2RandomLetters(),
      });
    }
  }, [gameState?.answers]);

  return (
    <>
      <div className="counter">
        <div>
          <div className="correct">
            Правильно: {GameSelectors.getCorrectCount(gameState)}
          </div>
          <div className="incorrect">
            Неправильно: {GameSelectors.getIncorrectCount(gameState)}
          </div>
          <div className="average-response-time">
            Середній час відповіді:{" "}
            {GameSelectors.getAverageResponseTime(gameState)}c
          </div>
        </div>
      </div>
      <div>Яка з цих літер іде першою в алфавіті?</div>
      <div className="container">
        <div
          className="letter"
          onClick={() => dispatch(GameActions.addAnswer(gameState, 0))}
        >
          {LETTERS[gameState.selectedLettersIndexes[0]]}
        </div>
        <div
          className="letter"
          onClick={() => dispatch(GameActions.addAnswer(gameState, 1))}
        >
          {LETTERS[gameState.selectedLettersIndexes[1]]}
        </div>
      </div>
    </>
  );
};
