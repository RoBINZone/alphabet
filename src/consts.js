// action consts
export const CLEAR_GAME = "CLEAR_GAME";
export const SET_SCREEN = "SET_SCREEN";
export const SET_ANSWERS = "SET_ANSWERS";
export const ADD_ANSWER = "ADD_ANSWER";
export const SET_START_TIME = "SET_START_TIME";
export const SET_SELECTED_LETTERS_INDEXES = "SET_SELECTED_LETTERS_INDEXES";

export const WELCOME_SCREEN = "WELCOME_SCREEN";
export const GAMEPLAY_SCREEN = "GAMEPLAY_SCREEN";
export const STATISTICS_SCREEN = "STATISTICS_SCREEN";
export const LETTERS = "АБВГҐДЕЄЖЗИІЇЙКЛМНОПРСТУФХЦЧШЩЬЮЯ";
export const MAXTURNS = 20;
