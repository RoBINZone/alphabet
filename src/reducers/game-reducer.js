import {
  ADD_ANSWER, CLEAR_GAME,
  SET_ANSWERS,
  SET_SCREEN,
  SET_SELECTED_LETTERS_INDEXES,
  SET_START_TIME,
  WELCOME_SCREEN
} from "../consts.js";

export const reducer = (state, action) => {
  switch (action.type) {
    case CLEAR_GAME:
      return JSON.parse(JSON.stringify(initialState));
    case SET_SCREEN:
      return {
        ...state,
        screen: action.screen,
      };
    case SET_ANSWERS:
      return {
        ...state,
        answers: action.answers,
      };
    case ADD_ANSWER:
      return {
        ...state,
        answers: [...state.answers, action.answer],
      };
    case SET_SELECTED_LETTERS_INDEXES:
      return {
        ...state,
        selectedLettersIndexes: action.selectedLettersIndexes,
      };
    case SET_START_TIME:
      return {
        ...state,
        startTime: action.startTime,
      };
    default:
      return state;
  }
};

export const initialState = {
  screen: WELCOME_SCREEN,
  answers: [],
  selectedLettersIndexes: [],
};
