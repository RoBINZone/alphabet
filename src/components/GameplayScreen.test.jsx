import React from "react";
import { render } from "@testing-library/react";
import { expect, it, describe } from "vitest";
import { GameplayScreen } from "./GameplayScreen";
import { GAMEPLAY_SCREEN } from "../consts.js";
import { initialState } from "../reducers/game-reducer.js";

describe("Gameplay", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <GameplayScreen gameState={initialState} dispatch={() => undefined} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
