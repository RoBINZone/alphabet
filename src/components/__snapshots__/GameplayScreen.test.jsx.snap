// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gameplay > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <div>
      <div
        class="correct"
      >
        Правильно: 0
      </div>
      <div
        class="incorrect"
      >
        Неправильно: 0
      </div>
      <div
        class="average-response-time"
      >
        Середній час відповіді: 0c
      </div>
    </div>
  </div>
  <div>
    Яка з цих літер іде першою в алфавіті?
  </div>
  <div
    class="container"
  >
    <div
      class="letter"
    />
    <div
      class="letter"
    />
  </div>
</DocumentFragment>
`;

exports[`Gameplay > should match snapshot many answers 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <div>
      <div
        class="correct"
      >
        Правильно: 6
      </div>
      <div
        class="incorrect"
      >
        Неправильно: 3
      </div>
      <div
        class="average-response-time"
      >
        Середній час відповіді: 0.43c
      </div>
    </div>
  </div>
  <div>
    Яка з цих літер іде першою в алфавіті?
  </div>
  <div
    class="container"
  >
    <div
      class="letter"
    >
      М
    </div>
    <div
      class="letter"
    >
      Й
    </div>
  </div>
</DocumentFragment>
`;
